src/openslide-vendor-dicom.c

```diff
@@ -0,0 +1,943 @@
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+/*
+ * DICOM (.dcm) support
+ */
+
+#include "openslide-private.h"
+#include "openslide-decode-jpeg.h"
+#include "openslide-decode-sqlite.h"
+#include "openslide-hash.h"
+
+#include <glib.h>
+#include <glib-object.h>
+#include <gio/gio.h>
+#include <string.h>
+#include <errno.h>
+
+#include <dicom/dicom.h>
+
+#define FREEF(F, V) { \
+    if (V) { \
+        (F)((V)); \
+        (V) = NULL; \
+    } \
+}
+
+/* The SOP we check for.
+ */
+#define VLWholeSlideMicroscopyImageStorage "1.2.840.10008.5.1.4.1.1.77.1.6"
+
+/*
+#define DEBUG
+ */
+
+enum image_format {
+  FORMAT_UNKNOWN,
+  FORMAT_JPEG,
+};
+
+struct dicom_ops_data {
+  char *dirname;
+  int tile_size;
+};
+
+struct dicom_file {
+  char *filename;
+
+  GMutex lock;
+  DcmFilehandle *filehandle;
+  DcmDataSet *metadata;
+  DcmDataSet *meta;
+  DcmBOT *bot;
+};
+
+struct level {
+  struct _openslide_level base;
+  struct _openslide_grid *grid;
+
+  enum image_format image_format;
+  int image_width;
+  int image_height;
+  int tile_w;
+  int tile_h;
+  guint32 num_frames;
+  int tiles_across;
+  int tiles_down;
+
+  struct dicom_file *file;
+};
+
+struct associated {
+  struct _openslide_associated_image base;
+
+  const char *name;
+  struct dicom_file *file;
+};
+
+// a set of allowed image types for a class of image
+struct allowed_types { 
+  const char ***types;
+  int n_types;
+};
+
+// the ImageTypes we allow for pyr levels
+static const char *original_types[4] = {
+  "ORIGINAL", "PRIMARY", "VOLUME", "NONE"
+};
+static const char *resampled_types[4] = {
+  "DERIVED", "PRIMARY", "VOLUME", "RESAMPLED"
+};
+static const char **level_type_strings[] = {
+  original_types,
+  resampled_types
+};
+
+static const struct allowed_types level_types = {
+  level_type_strings, 
+  sizeof(level_type_strings) / sizeof(level_type_strings[0])
+};
+
+// the ImageTypes we allow for associated images
+static const char *label_types[4] = {
+  "ORIGINAL", "PRIMARY", "LABEL", "NONE",
+};
+static const char *overview_types[4] = {
+  "ORIGINAL", "PRIMARY", "OVERVIEW", "NONE",
+};
+static const char **associated_type_strings[] = {
+  label_types,
+  overview_types
+};
+static const struct allowed_types associated_types = {
+  associated_type_strings, 
+  sizeof(associated_type_strings) / sizeof(associated_type_strings[0])
+};
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(DcmFilehandle, dcm_filehandle_destroy)
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(DcmDataSet, dcm_dataset_destroy)
+
+static void set_gerror_from_dcm_error(GError **err, DcmError **dcm_error)
+{
+  char *msg = g_strdup_printf("libdicom %s: %s - %s",
+    dcm_error_code_str(dcm_error_code(*dcm_error)),
+    dcm_error_summary(*dcm_error),
+    dcm_error_message(*dcm_error));
+  g_set_error(err, OPENSLIDE_ERROR, OPENSLIDE_ERROR_FAILED, "%s", msg);
+  g_free(msg);
+  dcm_error_clear(dcm_error);
+}
+
+#ifdef DEBUG
+static void print_file(struct dicom_file *f) {
+  printf("file:\n" );
```

Comment:
It's very verbose right now (I like reading logs), I was planning to remove most of this stuff later in development when there's less basic machinery that needs verifying. Would that be OK?

Reply:
Removing them later is okay.  The codebase does already have a bunch of commented-out `g_debug()` calls in case they need to be re-enabled later (not my first choice of approach, but I inherited it), so I think it's okay to leave some of the debug prints in the codebase long-term.  To reduce the code verbosity, though, how about something like:

```c
#if 1
#define debug(...) g_debug(__VA_ARGS__)
#else
#define debug(...)
#endif
```

Reply:
Well, if it were me, I'd just call the debug functions unconditionally, and live with the small perf hit of iterating over some stuff only to call a debug function that's going to be compiled out.  As you say, this isn't long-term code.

Reply:
OK, I swapped it to just use `g_debug()` everywhere. It's much too chatty, but as you say, we can remove this before release.